Fix regex validation for empty string in non-required fields

jakobsjoqvist opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the documentation. (Should be made against the documentation branch)
- Create or update the tests.
- Refer to the issue you are closing in the PR description - fix #issue
- Specify if the PR is in WIP (work in progress) state or ready to be merged

Please ensure you read through the Contributing Guide: https://github.com/strapi/strapi/blob/master/CONTRIBUTING.md
-->

### What does it do?

Fixes a validation bug that caused empty strings to not validate in non-required fields with a regex constraint. This PR contains fixes for both frontend (admin) and backend validation.

### Why is it needed?

It should be possible to leave out non-required fields even if they uses regex validation.

### How to test it?

1. Go to content-type builder and create a text field that is not required but has some form of RegExp validation (e.g., `^\d+$`)
2. Go to content-manager and create a new entry with that field correctly filled in (as a number)
3. Edit that same entry and remove that field's value (back to blank).
4. Update should be successful.
5. Update the content type and set the regex field to be required.
6. Update the entry created in step 2 and try to save it with the regex field empty (update should fail)
7. Fill in a valid value and select update, update should now be successfull.

### Related issue(s)/PR(s)

Related issue: #11781

### PR State
PR is ready to be merged.

Elhebert commented on packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/utils/schema.js:
```diff
@@ -338,7 +338,10 @@ const createYupSchemaAttribute = (type, validations, options) => {
-          schema = schema.matches(new RegExp(validationValue), errorsTrads.regex);
+          schema = schema.matches(new RegExp(validationValue), {
+            message: errorsTrads.regex,
+            excludeEmptyString: validations.required !== true,
```
```suggestion
            excludeEmptyString: !validations.required,
```

Elhebert commented on packages/core/strapi/lib/services/entity-validator/validators.js:
```diff
@@ -104,7 +104,9 @@ const addMaxFloatValidator = (validator, { attr }) =>
-  _.isUndefined(attr.regex) ? validator : validator.matches(new RegExp(attr.regex));
+  _.isUndefined(attr.regex)
+    ? validator
+    : validator.matches(new RegExp(attr.regex), { excludeEmptyString: attr.required !== true });
```
```suggestion
    : validator.matches(new RegExp(attr.regex), { excludeEmptyString: !attr.required });
```

jakobsjoqvist replied:
This is fixed in commit #5022f38.

Elhebert reviewed (changes requested):
Thanks for the PR! It looks good to me.

As `required` is a boolean (correct me if I'm wrong @alexandrebodin) you can do `!required` instead of `required !== true`. It's a bit less wordy.

It would be perfect if we had a test to support this feature (at least in the back-end). We already have a test suite for the validators. It's focused on the `unique` validator, but it can be expanded to include a test for the `regexp` validator.

jakobsjoqvist commented:
> Thanks for the PR! It looks good to me.
> 
> As `required` is a boolean (correct me if I'm wrong @alexandrebodin) you can do `!required` instead of `required !== true`. It's a bit less wordy.
> 
> It would be perfect if we had a test to support this feature (at least in the back-end). We already have a test suite for the validators. It's focused on the `unique` validator, but it can be expanded to include a test for the `regexp` validator.

Thanks for the comments! I will try to add some tests during the weekend!

jakobsjoqvist commented:
Once again, thanks for the feedback! I have now addressed your comments and added tests for the backend, please let me know if you want me to fix anything else :)

Elhebert commented on packages/core/strapi/lib/services/entity-validator/__tests__/validators.test.js:
```diff
@@ -20,6 +20,102 @@ describe('Entity validator', () => {
     fakeFindOne.mockReset();
   });
 
+  describe('String RegExp validaor', () => {
```
```suggestion
  describe('String RegExp validator', () => {
```

Elhebert commented on packages/core/strapi/lib/services/entity-validator/__tests__/validators.test.js:
```diff
@@ -20,6 +20,102 @@ describe('Entity validator', () => {
+      },
+    };
+
+    test('Empty string in required field', () => {
```
Can you use a more descriptive test name?

Something like "_It fails the validation of an empty string for a required field_" so that we know exactly what the test does without having to look at the content.

Elhebert commented on packages/core/strapi/lib/services/entity-validator/__tests__/validators.test.js:
```diff
@@ -20,6 +20,102 @@ describe('Entity validator', () => {
+      return expect(validator('')).rejects.toBeInstanceOf(YupValidationError);
+    });
+
+    test('Valid regex in required field', () => {
```
Same as above.

Elhebert commented on packages/core/strapi/lib/services/entity-validator/__tests__/validators.test.js:
```diff
@@ -20,6 +20,102 @@ describe('Entity validator', () => {
+      return expect(validator(value)).resolves.toEqual(value);
+    });
+
+    test('Empty string in not required field', async () => {
```
Same as above.

Elhebert commented on packages/core/strapi/lib/services/entity-validator/__tests__/validators.test.js:
```diff
@@ -20,6 +20,102 @@ describe('Entity validator', () => {
+      return expect(validator(value)).resolves.toEqual(value);
+    });
+
+    test('Valid regex in not required field', () => {
```
Same as above.

Elhebert commented on packages/core/strapi/lib/services/entity-validator/__tests__/validators.test.js:
```diff
@@ -20,6 +20,102 @@ describe('Entity validator', () => {
+      privateAttributes: [],
+      options: {},
+      attributes: {
+        attrStringRequiredRegex: { type: 'string', required: false },
```
```suggestion
        attrStringRequiredRegex: { type: 'string', required: true },
```

Elhebert reviewed (changes requested):
Thanks for the tests, looking good 👌 
I added some suggestions to make the tests clearer.

jakobsjoqvist commented:
Thanks for the feedback @Elhebert! I have addressed your comments :) Please have a look and let me know if I should fix anything.

Elhebert reviewed (approved):
LGTM 👍
